File: examples/typescript-client/src/examples.ts
/**
 * Example usage scenarios demonstrating type-safe API interactions
 */

import { LightEngineClient } from './client';
import type {
  SensorPayload,
  NetworkTestRequest,
  DeviceCommandRequest,
  FailsafePowerRequest,
  AutomationRule,
} from './types';

async function runExamples() {
  // Initialize client
  const client = new LightEngineClient({
    baseUrl: 'http://localhost:8000',
    timeout: 10000,
  });

  console.log('🌱 Light Engine Charlie - Type-Safe API Examples\n');

  try {
    // Example 1: Health Check
    console.log('1️⃣ Health Check');
    const health = await client.health();
    console.log(`   Status: ${health.status}`);
    console.log(`   Version: ${health.version}\n`);

    // Example 2: Ingest Sensor Data (Type-Safe)
    console.log('2️⃣ Ingest Sensor Data (Type-Safe)');
    const sensorData: SensorPayload = {
      scope: 'VegRoom1',
      ts: new Date().toISOString(),
      sensors: {
        temperature: { value: 75.2, unit: 'F' },
        humidity: { value: 60.5, unit: '%' },
        co2: { value: 1200, unit: 'ppm' },
      },
    };
    const ingestResult = await client.ingestSensorData(sensorData);
    console.log(`   Ingested: ${JSON.stringify(ingestResult.message)}\n`);

    // Example 3: Get Latest Readings
    console.log('3️⃣ Get Latest Readings');
    const latest = await client.getLatestReadings('VegRoom1');
    console.log(`   Scope: ${latest.scope}`);
    console.log(`   Sensors: ${Object.keys(latest.sensors).join(', ')}`);
    console.log(`   Observed: ${latest.observedAt}\n`);

    // Example 4: Trigger Device Discovery
    console.log('4️⃣ Trigger Device Discovery');
    const discoveryResult = await client.triggerDiscovery();
    console.log(`   Status: ${discoveryResult.status}`);
    console.log(`   Message: ${discoveryResult.message}\n`);

    // Wait for discovery
    await new Promise((resolve) => setTimeout(resolve, 2000));

    // Example 5: Get Discovered Devices (with type safety)
    console.log('5️⃣ Get Discovered Devices');
    const devices = await client.getDiscoveredDevices();
    console.log(`   Found: ${devices.count} devices`);
    devices.devices.forEach((device) => {
      console.log(`   - ${device.name} (${device.protocol})`);
    });
    console.log();

    // Example 6: Network Connectivity Test (Type-Safe Request)
    console.log('6️⃣ Network Connectivity Test');
    const networkTest: NetworkTestRequest = {
      host: 'google.com',
      port: 80,
      protocol: 'http',
    };
    const networkResult = await client.testNetworkConnection(networkTest);
    console.log(`   Host: ${networkResult.host}:${networkResult.port}`);
    console.log(`   Reachable: ${networkResult.reachable ? '✅' : '❌'}`);
    console.log(`   Message: ${networkResult.message}\n`);

    // Example 7: Device Command (Type-Safe)
    console.log('7️⃣ Send Device Command');
    const command: DeviceCommandRequest = {
      device_id: 'fixture_001',
      command: {
        action: 'set_brightness',
        value: 80,
      },
    };
    const commandResult = await client.sendDeviceCommand(command);
    console.log(`   Device: ${commandResult.device_id}`);
    console.log(`   Success: ${commandResult.success ? '✅' : '❌'}`);
    console.log(`   Message: ${commandResult.message}\n`);

    // Example 8: Get Lighting Fixtures
    console.log('8️⃣ Get Lighting Fixtures');
    const fixtures = await client.getLightingFixtures();
    console.log(`   Total: ${fixtures.count} fixtures`);
    fixtures.fixtures.forEach((fixture) => {
      console.log(`   - ${fixture.name} (${fixture.protocol})`);
    });
    console.log();

    // Example 9: Create Automation Rule (Type-Safe)
    console.log('9️⃣ Create Automation Rule');
    const rule: AutomationRule = {
      name: 'High Temperature Alert',
      enabled: true,
      conditions: {
        sensor: 'temperature',
        operator: 'gt',
        value: 85,
        scope: 'VegRoom1',
      },
      actions: {
        notification: {
          type: 'alert',
          message: 'Temperature exceeded 85°F',
        },
        device_command: {
          device_id: 'exhaust_fan_001',
          action: 'turn_on',
        },
      },
      priority: 10,
    };
    const ruleResult = await client.createRule(rule);
    console.log(`   Rule ID: ${ruleResult.rule_id}`);
    console.log(`   Success: ${ruleResult.success ? '✅' : '❌'}\n`);

    // Example 10: List All Rules
    console.log('🔟 List Automation Rules');
    const rules = await client.listRules();
    console.log(`   Total: ${rules.count} rules`);
    rules.rules.forEach((r) => {
      console.log(`   - ${r.name} (${r.enabled ? 'enabled' : 'disabled'})`);
    });
    console.log();

    // Example 11: Emergency Failsafe (Type-Safe)
    console.log('1️⃣1️⃣ Emergency Failsafe Test');
    const failsafe: FailsafePowerRequest = {
      fixtures: ['fixture_001', 'fixture_002'],
      power: 'off',
      brightness: 0,
    };
    const failsafeResult = await client.lightingFailsafe(failsafe);
    console.log(`   Total: ${failsafeResult.total} fixtures`);
    console.log(`   Successful: ${failsafeResult.successful}/${failsafeResult.total}`);
    console.log();

    console.log('✅ All examples completed successfully!');
    console.log('\n📊 Type Safety Benefits:');
    console.log('   - IDE autocomplete for all API methods');
    console.log('   - Compile-time type checking');
    console.log('   - IntelliSense for request/response objects');
    console.log('   - Catch errors before runtime');

  } catch (error) {
    console.error('❌ Error:', error instanceof Error ? error.message : error);
    process.exit(1);
  }
}

// Run examples if this file is executed directly
if (require.main === module) {
  runExamples();
}

export { runExamples };

File: examples/python-client/examples.py
"""
Example usage scenarios for Light Engine Python SDK
Demonstrates both synchronous and asynchronous patterns
"""

import asyncio
from datetime import datetime
from light_engine import (
    LightEngineClient,
    AsyncLightEngineClient,
    SensorPayload,
    SensorReading,
    NetworkTestRequest,
    DeviceCommandRequest,
    FailsafePowerRequest,
    AutomationRule,
)


def synchronous_examples():
    """Synchronous API usage examples"""
    print("🌱 Light Engine Charlie - Python SDK Examples (Sync)\n")
    
    with LightEngineClient("http://localhost:8000") as client:
        # Example 1: Health Check
        print("1⃣ Health Check")
        health = client.health()
        print(f"   Status: {health.get('status')}")
        print(f"   Version: {health.get('version', 'N/A')}\n")
        
        # Example 2: Ingest Sensor Data
        print("2⃣ Ingest Sensor Data")
        payload = SensorPayload(
            scope="VegRoom1",
            ts=datetime.utcnow().isoformat() + "Z",
            sensors={
                "temperature": SensorReading(value=75.2, unit="F"),
                "humidity": SensorReading(value=60.5, unit="%"),
                "co2": SensorReading(value=1200, unit="ppm")
            }
        )
        result = client.ingest_sensor_data(payload)
        print(f"   Ingested: {result.get('success', True)}\n")
        
        # Example 3: Get Latest Readings
        print("3⃣ Get Latest Readings")
        latest = client.get_latest_readings("VegRoom1")
        print(f"   Scope: {latest['scope']}")
        print(f"   Sensors: {', '.join(latest['sensors'].keys())}\n")
        
        # Example 4: Trigger Discovery
        print("4⃣ Trigger Device Discovery")
        discovery = client.trigger_discovery()
        print(f"   Status: {discovery['status']}")
        print(f"   Message: {discovery['message']}\n")
        
        # Example 5: Network Test
        print("5⃣ Network Connectivity Test")
        net_test = NetworkTestRequest(host="google.com", port=80, protocol="http")
        net_result = client.test_network_connection(net_test)
        print(f"   Host: {net_result['host']}:{net_result['port']}")
        print(f"   Reachable: {'' if net_result['reachable'] else ''}\n")
        
        # Example 6: Device Command
        print("6⃣ Send Device Command")
        command = DeviceCommandRequest(
            device_id="fixture_001",
            command={"action": "set_brightness", "value": 80}
        )
        cmd_result = client.send_device_command(command)
        print(f"   Device: {cmd_result['device_id']}")
        print(f"   Success: {'' if cmd_result['success'] else ''}\n")
        
        # Example 7: Create Automation Rule
        print("7⃣ Create Automation Rule")
        rule = AutomationRule(
            name="High Temperature Alert",
            enabled=True,
            conditions={
                "sensor": "temperature",
                "operator": "gt",
                "value": 85
            },
            actions={
                "notification": {"type": "alert"},
                "device_command": {"device_id": "fan_001", "action": "turn_on"}
            },
            priority=10
        )
        rule_result = client.create_rule(rule)
        print(f"   Rule ID: {rule_result['rule_id']}")
        print(f"   Success: {'' if rule_result['success'] else ''}\n")
        
        # Example 8: Emergency Failsafe
        print("8⃣ Emergency Failsafe")
        failsafe = FailsafePowerRequest(
            fixtures=["fixture_001", "fixture_002"],
            power="off",
            brightness=0
        )
        failsafe_result = client.lighting_failsafe(failsafe)
        print(f"   Total: {failsafe_result['total']} fixtures")
        print(f"   Successful: {failsafe_result['successful']}/{failsafe_result['total']}\n")
        
        print(" All synchronous examples completed!\n")


async def asynchronous_examples():
    """Asynchronous API usage examples"""
    print("🌱 Light Engine Charlie - Python SDK Examples (Async)\n")
    
    async with AsyncLightEngineClient("http://localhost:8000") as client:
        # Example 1: Health Check
        print("1⃣ Health Check (Async)")
        health = await client.health()
        print(f"   Status: {health.get('status')}")
        print(f"   Version: {health.get('version', 'N/A')}\n")
        
        # Example 2: Parallel Requests
        print("2⃣ Parallel Requests (Async)")
        results = await asyncio.gather(
            client.get_scopes(),
            client.get_devices(),
            client.list_rules(),
            return_exceptions=True
        )
        print(f"   Scopes: {results[0].get('count', 0)}")
        print(f"   Devices: {results[1].get('count', 0)}")
        print(f"   Rules: {results[2].get('count', 0)}\n")
        
        # Example 3: Sensor History
        print("3⃣ Get Sensor History (Async)")
        history = await client.get_sensor_history("VegRoom1", "temperature", limit=20)
        print(f"   Scope: {history['scope']}")
        print(f"   Sensor: {history['sensor']}")
        print(f"   Data points: {len(history.get('history', []))}\n")
        
        # Example 4: Device Discovery
        print("4⃣ Device Discovery (Async)")
        await client.trigger_discovery()
        await asyncio.sleep(2)  # Wait for scan
        devices = await client.get_discovered_devices()
        print(f"   Discovered: {len(devices.get('devices', []))} devices\n")
        
        print(" All asynchronous examples completed!\n")


def error_handling_example():
    """Demonstrate error handling"""
    print("🛡 Error Handling Example\n")
    
    from light_engine.exceptions import APIError, TimeoutError
    
    client = LightEngineClient("http://localhost:8000", timeout=5)
    
    try:
        # This will work
        health = client.health()
        print(f" Health check succeeded: {health['status']}\n")
        
        # This might fail if scope doesn't exist
        try:
            latest = client.get_latest_readings("NonExistentScope")
        except APIError as e:
            print(f" Expected API error: {e}\n")
        
    except TimeoutError as e:
        print(f" Timeout: {e}\n")
    except APIError as e:
        print(f" API Error: {e} (Status: {e.status_code})\n")
    finally:
        client.session.close()


def type_hints_example():
    """Demonstrate type hints and IDE support"""
    print(" Type Hints Example\n")
    
    # Type hints help IDEs provide autocomplete
    client: LightEngineClient = LightEngineClient("http://localhost:8000")
    
    # IDE knows the return types
    health: dict = client.health()
    
    # Dataclasses provide structure
    payload: SensorPayload = SensorPayload(
        scope="TestRoom",
        ts=datetime.utcnow().isoformat() + "Z",
        sensors={
            "temp": SensorReading(value=72.0, unit="F")
        }
    )
    
    # Type checkers (mypy) can validate this
    result: dict = client.ingest_sensor_data(payload)
    
    print(" Type hints enable:")
    print("   - IDE autocomplete")
    print("   - Static type checking with mypy")
    print("   - Better documentation")
    print("   - Fewer runtime errors\n")
    
    client.session.close()


if __name__ == "__main__":
    print("=" * 60)
    print("Light Engine Charlie - Python SDK Examples")
    print("=" * 60 + "\n")
    
    # Run synchronous examples
    synchronous_examples()
    
    # Run asynchronous examples
    asyncio.run(asynchronous_examples())
    
    # Demonstrate error handling
    error_handling_example()
    
    # Show type hints benefits
    type_hints_example()
    
    print("=" * 60)
    print(" All examples completed successfully!")
    print("=" * 60)
